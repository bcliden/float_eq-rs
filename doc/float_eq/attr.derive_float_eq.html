<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Helper for deriving the various float_eq traits."><meta name="keywords" content="rust, rustlang, rust-lang, derive_float_eq"><title>derive_float_eq in float_eq - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings"></script><script src="../storage.js"></script><script src="../crates.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../favicon.svg">
<link rel="alternate icon" type="image/png" href="../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc attr"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../float_eq/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><div class="sidebar-elems"><h2 class="location">Other items in<br><a href="index.html">float_eq</a></h2><div id="sidebar-vars" data-name="derive_float_eq" data-ty="attr" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img src="../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img src="../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Attribute Macro <a href="index.html">float_eq</a>::<wbr><a class="attr" href="#">derive_float_eq</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></span></h1><pre class="rust attr">#[derive_float_eq]</pre><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Helper for deriving the various float_eq traits.</p>
<p>By default, this will derive <a href="trait.FloatEqUlpsTol.html"><code>FloatEqUlpsTol</code></a>, <a href="trait.FloatEq.html"><code>FloatEq</code></a>, <a href="trait.FloatEqDebugUlpsDiff.html"><code>FloatEqDebugUlpsDiff</code></a>
and <a href="trait.AssertFloatEq.html"><code>AssertFloatEq</code></a>.</p>
<p>Attribute parameters are passed through to the <code>#[float_eq(...)]</code> attribute:
<code>ulps_tol</code> is the name of the type generated for <a href="trait.FloatEqUlpsTol.html"><code>FloatEqUlpsTol</code></a> and
<code>debug_ulps_diff</code> the name of the type generated for <a href="trait.FloatEqDebugUlpsDiff.html"><code>FloatEqDebugUlpsDiff</code></a>.
Both are required.</p>
<p>If the optional <code>all_tol</code> parameter is provided, which names the type to be
used by <code>FloatEqAll::AllTol</code> then <a href="trait.FloatEqAll.html"><code>FloatEqAll</code></a> and <a href="trait.AssertFloatEqAll.html"><code>AssertFloatEqAll</code></a>
are also derived.</p>
<p>See <a href="https://jtempest.github.io/float_eq-rs/book/how_to/compare_custom_types.html#derive_float_eq">How to compare custom types</a> for more information and example usage.</p>
</div></details></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="float_eq" data-search-index-js="../search-index.js" data-search-js="../search.js"></div><script src="../main.js"></script></body></html>