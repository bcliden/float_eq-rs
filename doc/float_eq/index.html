<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `float_eq` crate."><meta name="keywords" content="rust, rustlang, rust-lang, float_eq"><title>float_eq - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings"></script><script src="../storage.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../favicon.svg">
<link rel="alternate icon" type="image/png" href="../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../float_eq/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><p class="location">Crate float_eq</p><div class="block version"><p>Version 0.6.0</p></div><div class="sidebar-elems"><a id="all-types" href="all.html"><p>See all float_eq's items</p></a><div class="block items"><ul><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Definitions</a></li></ul></div><p class="location"></p><div id="sidebar-vars" data-name="float_eq" data-ty="mod" data-relpath="../"></div></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" class="help-button">?</button>
                <a id="settings-menu" href="../settings.html"><img src="../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="">float_eq</a></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/float_eq/lib.rs.html#1-104" title="goto source code">[src]</a></span></h1><div class="docblock"><p>Compare IEEE floating point primitives, structs and collections for equality.</p>
<p>This is API reference documentation. For introductory material, guides and
discussion see <a href="https://jtempest.github.io/float_eq-rs/book/index.html">the float_eq book</a>.</p>
<h1 id="basic-usage" class="section-header"><a href="#basic-usage">Basic usage</a></h1>
<p>This crate provides boolean comparisons via <a href="../float_eq/macro.float_eq.html" title="float_eq!"><code>float_eq!</code></a> and <a href="../float_eq/macro.float_ne.html" title="float_ne!"><code>float_ne!</code></a>:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">float_eq</span>::<span class="ident">float_eq</span>;

<span class="kw">if</span> (<span class="macro">float_eq</span><span class="macro">!</span>(<span class="ident">y_pos</span>, <span class="number">0.0</span>, <span class="ident">abs</span> <span class="op">&lt;</span><span class="op">=</span> <span class="number">0.000_1</span>)) {
   <span class="comment">//...</span>
}</pre></div>
<p>And asserts via <a href="../float_eq/macro.assert_float_eq.html" title="assert_float_eq!"><code>assert_float_eq!</code></a> and <a href="../float_eq/macro.assert_float_ne.html" title="assert_float_ne!"><code>assert_float_ne!</code></a>:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">float_eq</span>::<span class="ident">assert_float_eq</span>;

<span class="kw">const</span> <span class="ident">TOL</span>: <span class="ident">f32</span> <span class="op">=</span> <span class="number">0.000_366_210_94</span>;
<span class="macro">assert_float_eq</span><span class="macro">!</span>(<span class="number">0.1f32</span>.<span class="ident">recip</span>(), <span class="number">10.0</span>, <span class="ident">r2nd</span> <span class="op">&lt;</span><span class="op">=</span> <span class="ident">TOL</span>);</pre></div>
<p>Each of which invokes a specific comparison algorithm with an explictly
provided toelrance. In these examples:</p>
<ul>
<li><code>abs &lt;= 0.000_1</code> is an absolute tolerance comparison with a tolerance of <code>0.000_1</code>.</li>
<li><code>r2nd &lt;= TOL</code> is a relative tolerance comparison with a tolerance of <code>TOL</code>,
scaled to the precision of the second operand.</li>
</ul>
<h1 id="comparison-algorithms" class="section-header"><a href="#comparison-algorithms">Comparison algorithms</a></h1>
<p>These are always of the form <code>CHECK &lt;= tol</code>, where <code>CHECK</code> is one of:</p>
<ul>
<li><code>abs</code>: an <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#absolute-tolerance-comparison">absolute tolerance comparison</a>.</li>
<li><code>rmax</code>: a <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#relative-tolerance-comparison">relative tolerance comparison</a>, scaled to the precision of the larger operand/field.</li>
<li><code>rmin</code>: a <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#relative-tolerance-comparison">relative tolerance comparison</a>, scaled to the precision of the smaller operand/field.</li>
<li><code>r1st</code>: a <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#relative-tolerance-comparison">relative tolerance comparison</a>, scaled to the precision of the first operand/field.</li>
<li><code>r2nd</code>: a <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#relative-tolerance-comparison">relative tolerance comparison</a>, scaled to the precision of the second operand/field.</li>
<li><code>ulps</code>: an <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#units-in-the-last-place-ulps-comparison">ULPs comparison</a>.</li>
</ul>
<p>When comparing homogeneous composite types that implement <a href="../float_eq/trait.FloatEqAll.html" title="FloatEqAll"><code>FloatEqAll</code></a>,
variants that use a uniform <code>tol</code> across all fields are also available:</p>
<ul>
<li><code>abs_all</code>: an <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#absolute-tolerance-comparison">absolute tolerance comparison</a>.</li>
<li><code>rmax_all</code>: a <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#relative-tolerance-comparison">relative tolerance comparison</a>, scaled to the precision of the larger field.</li>
<li><code>rmin_all</code>: a <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#relative-tolerance-comparison">relative tolerance comparison</a>, scaled to the precision of the smaller field.</li>
<li><code>r1st_all</code>: a <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#relative-tolerance-comparison">relative tolerance comparison</a>, scaled to the precision of the first field.</li>
<li><code>r2nd_all</code>: a <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#relative-tolerance-comparison">relative tolerance comparison</a>, scaled to the precision of the second field.</li>
<li><code>ulps_all</code>: an <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#units-in-the-last-place-ulps-comparison">ULPs comparison</a>.</li>
</ul>
<p><em>Note: <code>rel</code> and <code>rel_all</code> are legacy aliases for <code>rmax</code> and <code>rmax_all</code>, but
using the more precise algorithm names is recommended.</em></p>
<h1 id="combining-checks" class="section-header"><a href="#combining-checks">Combining checks</a></h1>
<p>If multiple checks are specified in either a boolean comparison or an assert,
they are applied left to right and will shortcut on success. For example:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="macro">float_eq</span><span class="macro">!</span>(<span class="ident">a</span>, <span class="ident">b</span>, <span class="ident">abs</span> <span class="op">&lt;</span><span class="op">=</span> <span class="ident">abs_tol</span>, <span class="ident">ulps</span> <span class="op">&lt;</span><span class="op">=</span> <span class="ident">ulps_tol</span>)</pre></div>
<p>Is equivalent to:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="macro">float_eq</span><span class="macro">!</span>(<span class="ident">a</span>, <span class="ident">b</span>, <span class="ident">abs</span> <span class="op">&lt;</span><span class="op">=</span> <span class="ident">abs_tol</span>) <span class="op">|</span><span class="op">|</span> <span class="macro">float_eq</span><span class="macro">!</span>(<span class="ident">a</span>, <span class="ident">b</span>, <span class="ident">ulps</span> <span class="op">&lt;</span><span class="op">=</span> <span class="ident">ulps_tol</span>)</pre></div>
<h1 id="extending-float_eq-over-custom-types" class="section-header"><a href="#extending-float_eq-over-custom-types">Extending float_eq over custom types</a></h1>
<p>See <a href="https://jtempest.github.io/float_eq-rs/book/how_to/compare_custom_types.html">How to compare custom types</a>.</p>
</div><h2 id="macros" class="section-header"><a href="#macros">Macros</a></h2>
<table><tr class="module-item"><td><a class="macro" href="macro.assert_float_eq.html" title="float_eq::assert_float_eq macro">assert_float_eq</a></td><td class="docblock-short"><p>Asserts that two floating point expressions are equal to each other.</p>
</td></tr><tr class="module-item"><td><a class="macro" href="macro.assert_float_ne.html" title="float_eq::assert_float_ne macro">assert_float_ne</a></td><td class="docblock-short"><p>Asserts that two floating point expressions are not equal to each other.</p>
</td></tr><tr class="module-item"><td><a class="macro" href="macro.debug_assert_float_eq.html" title="float_eq::debug_assert_float_eq macro">debug_assert_float_eq</a></td><td class="docblock-short"><p>Asserts that two floating point expressions are equal to each other.</p>
</td></tr><tr class="module-item"><td><a class="macro" href="macro.debug_assert_float_ne.html" title="float_eq::debug_assert_float_ne macro">debug_assert_float_ne</a></td><td class="docblock-short"><p>Asserts that two floating point expressions are not equal to each other.</p>
</td></tr><tr class="module-item"><td><a class="macro" href="macro.float_eq.html" title="float_eq::float_eq macro">float_eq</a></td><td class="docblock-short"><p>Checks if two floating point expressions are equal to each other.</p>
</td></tr><tr class="module-item"><td><a class="macro" href="macro.float_ne.html" title="float_eq::float_ne macro">float_ne</a></td><td class="docblock-short"><p>Checks if two floating point expressions are not equal to each other.</p>
</td></tr></table><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<table><tr class="module-item"><td><a class="struct" href="struct.ComplexUlps.html" title="float_eq::ComplexUlps struct">ComplexUlps</a></td><td class="docblock-short"><p>The absolute difference between two floating point <a href="https://docs.rs/num/0.3.0/num/struct.Complex.html"><code>num::Complex&lt;T&gt;</code></a> instances
in ULPs.</p>
</td></tr></table><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<table><tr class="module-item"><td><a class="trait" href="trait.AssertFloatEq.html" title="float_eq::AssertFloatEq trait">AssertFloatEq</a></td><td class="docblock-short"><p>Debug context for when an assert fails.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.AssertFloatEqAll.html" title="float_eq::AssertFloatEqAll trait">AssertFloatEqAll</a></td><td class="docblock-short"><p>Debug context for when an assert using an <code>all</code> check fails.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.FloatEq.html" title="float_eq::FloatEq trait">FloatEq</a></td><td class="docblock-short"><p>Compare IEEE floating point values for equality using per-field tolerances.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.FloatEqAll.html" title="float_eq::FloatEqAll trait">FloatEqAll</a></td><td class="docblock-short"><p>Compare IEEE floating point values for equality using a uniform tolerance.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.FloatEqDebugUlpsDiff.html" title="float_eq::FloatEqDebugUlpsDiff trait">FloatEqDebugUlpsDiff</a></td><td class="docblock-short"><p>Per-field results of <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#units-in-the-last-place-ulps-comparison">ULPs</a>
based diff calculations.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.FloatEqUlpsTol.html" title="float_eq::FloatEqUlpsTol trait">FloatEqUlpsTol</a></td><td class="docblock-short"><p>Per-field tolerances for <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#units-in-the-last-place-ulps-comparison">ULPs comparisons</a>.</p>
</td></tr></table><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<table><tr class="module-item"><td><a class="type" href="type.ComplexUlps32.html" title="float_eq::ComplexUlps32 type">ComplexUlps32</a></td><td class="docblock-short"><p><a href="../float_eq/struct.ComplexUlps.html" title="ComplexUlps&lt;T&gt;"><code>ComplexUlps&lt;T&gt;</code></a> type matching <a href="https://docs.rs/num-complex/0.3/num_complex/type.Complex32.html"><code>num::Complex32</code></a>.</p>
</td></tr><tr class="module-item"><td><a class="type" href="type.ComplexUlps64.html" title="float_eq::ComplexUlps64 type">ComplexUlps64</a></td><td class="docblock-short"><p><a href="../float_eq/struct.ComplexUlps.html" title="ComplexUlps&lt;T&gt;"><code>ComplexUlps&lt;T&gt;</code></a> type matching <a href="https://docs.rs/num-complex/0.3/num_complex/type.Complex64.html"><code>num::Complex64</code></a>.</p>
</td></tr><tr class="module-item"><td><a class="type" href="type.DebugUlpsDiff.html" title="float_eq::DebugUlpsDiff type">DebugUlpsDiff</a></td><td class="docblock-short"><p>Per-field results of <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#units-in-the-last-place-ulps-comparison">ULPs</a> based diff calculations.</p>
</td></tr><tr class="module-item"><td><a class="type" href="type.UlpsTol.html" title="float_eq::UlpsTol type">UlpsTol</a></td><td class="docblock-short"><p>Per-field tolerances for <a href="https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html#units-in-the-last-place-ulps-comparison">ULPs comparisons</a>.</p>
</td></tr></table><h2 id="attributes" class="section-header"><a href="#attributes">Attribute Macros</a></h2>
<table><tr class="module-item"><td><a class="attr" href="attr.derive_float_eq.html" title="float_eq::derive_float_eq attr">derive_float_eq</a></td><td class="docblock-short"><p>Helper for deriving the various float_eq traits.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="float_eq"></div>
    <script src="../main.js"></script><script defer src="../search-index.js"></script></body></html>