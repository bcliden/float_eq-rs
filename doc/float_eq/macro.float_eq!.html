<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.float_eq.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.float_eq.html">macro.float_eq.html</a>...</p>
    <script>location.replace("macro.float_eq.html" + location.search + location.hash);</script>
</body>
</html>