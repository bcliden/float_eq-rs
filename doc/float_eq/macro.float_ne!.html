<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.float_ne.html">
</head>
<body>
    <p>Redirecting to <a href="macro.float_ne.html">macro.float_ne.html</a>...</p>
    <script>location.replace("macro.float_ne.html" + location.search + location.hash);</script>
</body>
</html>