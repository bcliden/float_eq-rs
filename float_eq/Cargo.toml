[package]
name = "float_eq"
version = "0.5.0"
authors = ["jtempest"]
license = "MIT OR Apache-2.0"
description = "Compare IEEE floating point values for equality."
homepage = "https://github.com/jtempest/float_eq-rs"
repository = "https://github.com/jtempest/float_eq-rs"
documentation = "https://docs.rs/float_eq/"
keywords = ["approximate", "assert", "comparison", "equality", "float"]
categories = ["algorithms", "development-tools::debugging", "no-std"]
readme = "crates-io.md"
include = ["Cargo.toml", "src/**/*.rs", "crates-io.md", "LICENSE-APACHE", "LICENSE-MIT"]
edition = "2018"

[package.metadata.docs.rs]
all-features = true

[badges]
maintenance = { status = "experimental" }
travis-ci = { repository = "jtempest/float_eq-rs" }
# Turned off for now since coveralls has better support for testing multiple 
# feature sets, and thus is more representative.
# codecov = { repository = "jtempest/float_eq-rs" }
coveralls = { repository = "jtempest/float_eq-rs" }

[dev-dependencies]
trybuild = "1"

[dependencies.num-complex]
version = "0.3"
optional = true

[dependencies.float_eq_derive]
version = "=0.5.0"
optional = true
path = "../float_eq_derive"

[features]
default = ["std"]
std = []
num = ["num-complex"]
derive = ["float_eq_derive"]