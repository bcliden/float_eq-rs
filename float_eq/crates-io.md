# float_eq

[![crate](https://img.shields.io/crates/v/float_eq.svg)](https://crates.io/crates/float_eq)
[![documentation](https://docs.rs/float_eq/badge.svg)](https://docs.rs/float_eq)
[![Travis status](https://travis-ci.com/jtempest/float_eq-rs.svg?branch=master)](https://travis-ci.com/github/jtempest/float_eq-rs)
[![Coverage Status](https://coveralls.io/repos/github/jtempest/float_eq-rs/badge.svg?branch=master)](https://coveralls.io/github/jtempest/float_eq-rs?branch=master)

Compare IEEE floating point values for equality.

Comparing floating point values for equality is notoriously difficult,
getting it right requires careful reasoning and iteration. This API provides
a variety of comparison algorithms and debugging tools to help make the
process more intuitive and your choices explicit and clear to future
maintainers.

# Background

Given how widely algorithmic requirements can vary, `float_eq` explores the
idea that there are no generally sensible default margins for comparisons.
This is in contrast to the approach taken by many other crates, which often
provide default epsilon values in checks or implicitly favour particular
algorithms. The author's hope is that by exposing the inherent complexity
in a uniform way, programmers will find it easier to develop an intuition
for effective use of floats.

This work builds on the definitions in Knuth's The Art Of Computer Programming,
(Vol. 2, Seminumerical Algorithms, Third Edition, section 4.2.2), and *that* 
Random ASCII article on [floating point comparison].

## Usage

Add this to your cargo.toml:

```
[dependencies]
float_eq = "0.5"
```

and, if you're using the 2015 edition, this to your crate root:

```rust
extern crate float_eq;
```

then, you can import items with `use`:

```rust
use float_eq::{assert_float_eq, float_eq};
```

## Comparisons

This crate provides boolean comparison operations:

```rust
assert!(float_eq!(1000.0f32, 1000.0002, ulps <= 4));

const ROUNDING_ERROR: f32 = 0.000_345_266_98;
assert!(float_ne!(4.0f32, 4.1, rmax <= ROUNDING_ERROR));
```

And asserts:

```rust
const RECIP_REL_EPSILON: f32 = 0.000_366_210_94; 
assert_float_eq!(0.1f32.recip(), 10.0, r2nd <= RECIP_REL_EPSILON);

assert_float_ne!(0.0f32, 0.000_1, abs <= 0.000_05, ulps <= 4);
```

Checks are invoked by name and with a threshold, so for example `abs <= 0.000_05`
should be read as *"an absolute epsilon comparison with a maximum difference of
less than or equal to `0.000_05`"*. Similarly, `rmax`, `rmin`, `r1st` and `r2nd`
provide a variety of kinds of relative epsilon comparison with thresholds that
scale to the granularity of one or input value or the other and `ulps` is an
ULPs based comparison that takes advantage of the underlying bitwise 
representation. See the [API documentation] for a long form introduction
to the different kinds of checks, their uses and limitations.

## Combining checks

If more than one check is specified by a comparison then they are performed
in order from left to right. If any check is true, then the two values are
considered equal. For example, this expression:

```rust
float_eq!(a, b, abs <= 0.000_01, ulps <= 4)
```

Is equivalent to:

```rust
float_eq!(a, b, abs <= 0.000_01) || float_eq!(a, b, ulps <= 4)
```

This allows you to build comparison expressions as needed, only paying for what
you use.

## Composite types

Composite types that implement `FloatEq` may be compared on a field-by-field
basis, and types that implement `FloatEqAll` may be compared with a uniformly
applied epsilon value across all fields:

```rust
let a = Complex32 { re: 2.0, im: 4.000_002 };
let b = Complex32 { re: 2.000_000_5, im: 4.0 };

assert_float_eq!(a, b, ulps <= Complex32Ulps { re: 2, im: 4 });
assert_float_eq!(a, b, ulps_all <= 4);
```

Arrays of size 0 to 32 (inclusive) are supported:

```rust
let a = [1.0, -2.0, 3.0];
let b = [-1.0, 2.0, 3.5];
assert_float_eq!(a, b, abs <= [2.0, 4.0, 0.5]);
assert_float_eq!(a, b, abs_all <= 4.0);
```

As are tuples up to size 12 (inclusive):

```rust
let a = (1.0f32, 2.0f64);
let b = (1.5f32, -2.0f64);
assert_float_eq!(a, b, abs <= (0.5, 4.0));
```

There are also blanket trait impls for comparing mutable and immutable reference
types, the contents of `Cell`, `RefCell`, `Rc`, `Arc` and `Box` instances, as
well as for slices, `Option`, `Vec`, `VecDeque`, `LinkedList`, `BTreeMap` and
`HashMap`.

Types that also implement `AssertFloatEq`/`AssertFloatEqAll` may be used in the
assert forms.

## Derivable

If the optional `"derive"` feature is enabled, all of the traits may be 
implemented using `#[derive]`. The easiest way to do so is to make use of the 
`#[derive_float_eq]` helper macro:

```rust
#[derive_float_eq(
    ulps_epsilon = "PointUlps",
    debug_ulps_diff = "PointUlpsDebugUlpsDiff",
    all_epsilon = "f64"
)]
#[derive(Debug, PartialEq, Clone, Copy)]
struct Point {
    x: f64,
    y: f64,
}

let a = Point { x: 1.0, y: -2.0 };
let c = Point { 
    x: 1.000_000_000_000_000_9, 
    y: -2.000_000_000_000_001_3
};
assert_float_eq!(a, c, ulps <= PointUlps { x: 4, y: 3 });
assert_float_eq!(a, c, ulps_all <= 4);
```

## Error messages

Assertion failure output tries to provide useful context information without
going overboard. For example, this call:

```rust
assert_float_eq!(4.0f32, 4.000_008, rmax <= 0.000_001);
```

Panics with this error message:

```
thread 'main' panicked at 'assertion failed: `float_eq!(left, right, rmax <= ε)`
        left: `4.0`,
       right: `4.000008`,
    abs_diff: `0.000008106232`,
   ulps_diff: `Some(17)`,
    [rmax] ε: `0.000004000008`', assert_failure.rs:15:5
```

The message shows the values of the expressions being compared and the
difference between them both in absolute terms and in terms of ULPs. The 
`[rmax] ε` line shows the epsilon value that the absolute difference was
compared against after being appropriately scaled.

## Optional features

This crate can be used without the standard library (`#![no_std]`) by disabling
the default `std` feature. Use this in `Cargo.toml`:

```
[dependencies.float_eq]
version = "0.5"
default-features = false
```

Other optional features:
- **derive** — provides custom derive macros for all traits.
- **num** — blanket trait impls for `num::Complex` where it is instanced with a
  compatible type.

## Related efforts

The [`approx`] and [`float-cmp`] crates provide a similar style of general
comparison operations, whereas [`assert_float_eq`] focuses specifically on
assertions. The [`almost`] crate instead divides its API into algorithms 
comparing against zero and non-zero values. In contrast, [`efloat`] takes the
approach of tracking the error bounds of values as operations are applied.

## Contributing 

Constructive feedback, suggestions and contributions welcomed, please
[open an issue].

## Changelog

Release information is available in [CHANGELOG.md](CHANGELOG.md).

[API documentation]: https://docs.rs/float_eq
[floating point comparison]: https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/
[open an issue]: https://github.com/jtempest/float_eq-rs/issues/
[`almost`]: https://crates.io/crates/almost
[`approx`]: https://crates.io/crates/approx
[`assert_float_eq`]: https://crates.io/crates/assert_float_eq
[`efloat`]: https://crates.io/crates/efloat
[`float-cmp`]: https://crates.io/crates/float-cmp
