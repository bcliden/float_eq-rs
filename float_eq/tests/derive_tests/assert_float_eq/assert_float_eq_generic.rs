use float_eq::{AssertFloatEq, FloatEq, FloatEqDebugUlpsDiff, FloatEqUlpsTol};

#[derive(
    Debug, Clone, Copy, PartialEq, FloatEqUlpsTol, FloatEq, FloatEqDebugUlpsDiff, AssertFloatEq,
)]
struct MyComplexUlps<T> {
    re: T,
    im: T,
}

fn main() {}
