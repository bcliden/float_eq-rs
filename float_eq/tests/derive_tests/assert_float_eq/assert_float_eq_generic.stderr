error: This trait does not yet support derive for generic types.
 --> $DIR/assert_float_eq_generic.rs:4:36
  |
4 |     Debug, Clone, Copy, PartialEq, FloatEqUlpsTol, FloatEq, FloatEqDebugUlpsDiff, AssertFloatEq,
  |                                    ^^^^^^^^^^^^^^
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: This trait does not yet support derive for generic types.
 --> $DIR/assert_float_eq_generic.rs:4:52
  |
4 |     Debug, Clone, Copy, PartialEq, FloatEqUlpsTol, FloatEq, FloatEqDebugUlpsDiff, AssertFloatEq,
  |                                                    ^^^^^^^
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: This trait does not yet support derive for generic types.
 --> $DIR/assert_float_eq_generic.rs:4:61
  |
4 |     Debug, Clone, Copy, PartialEq, FloatEqUlpsTol, FloatEq, FloatEqDebugUlpsDiff, AssertFloatEq,
  |                                                             ^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: This trait does not yet support derive for generic types.
 --> $DIR/assert_float_eq_generic.rs:4:83
  |
4 |     Debug, Clone, Copy, PartialEq, FloatEqUlpsTol, FloatEq, FloatEqDebugUlpsDiff, AssertFloatEq,
  |                                                                                   ^^^^^^^^^^^^^
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
